internal/cli/root.go

```diff
@@ -0,0 +1,82 @@
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published
+// by the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+//
+// You can be released from the requirements of the above licenses by purchasing
+// a commercial license. Buying such a license is mandatory if you want to
+// modify or otherwise use the software for commercial activities involving the
+// Arduino software without disclosing the source code of your own applications.
+// To purchase a commercial license, send an email to license@arduino.cc.
+
+// Package cli defines the command line interface.
+package cli
+
+import (
+	"os"
+
+	"github.com/arduino/libraries-repository-engine/internal/feedback"
+	"github.com/spf13/cobra"
+)
+
+// rootCmd defines the base CLI command.
+var rootCmd = &cobra.Command{
+	Use:  "libraries-repository-engine",
+	Long: "The tool for managing the Arduino Library Manager content.",
+	CompletionOptions: cobra.CompletionOptions{
+		DisableDefaultCmd: true,
+	},
+	Aliases: []string{"sync"},
+}
+
+func init() {
+	rootCmd.PersistentFlags().String("config-file", "config.json", "Configuration file path")
+}
+
+// Execute adds all child commands to the root command and sets flags appropriately.
+func Execute() error {
+	// Support the old `libraries-repository-engine [CONFIG_FILE [REGISTRY_FILE]]` interface.
+	func() {
+		if len(os.Args) > 1 {
+			if os.Args[1][0] == '-' {
+				// The argument is a flag, so assume the new interface.
+				return
+			}
+
+			for _, command := range rootCmd.Commands() {
+				for _, alias := range append(command.Aliases, command.Name(), "help") { // Hacky to check "help" redundantly, but whatever.
+					if os.Args[1] == alias {
+						// The argument is a registered subcommand, so assume the new interface.
+						return
+					}
+				}
+			}
+
+			// The argument is not a registered subcommand, so assume it is the CONFIG_FILE positional argument of the old interface.
+			rootCmd.PersistentFlags().Set("config-file", os.Args[1]) // Transfer the argument to the new interface's flag.
+			os.Args = append([]string{os.Args[0]}, os.Args[2:]...)   // Remove the argument.
+		}
+
+		feedback.Warning(
+			`Using deprecated command line syntax. New syntax:
+libraries-repository-engine sync [--config-file=CONFIG_FILE] [REGISTRY_FILE]
+`,
+		)
+		// Add the subcommand the root aliases.
+		os.Args = append([]string{os.Args[0], rootCmd.Aliases[0]}, os.Args[1:]...)
+	}()
```

Comment:
Am not 100% sure about it but I think this could be moved in a [`PreRun` hook](pre-run) in `rootCmd`. :thinking: 

[pre-run]: https://github.com/spf13/cobra/blob/master/user_guide.md#prerun-and-postrun-hooks

Reply:
I don't think it's possible. The CLI arguments are parsed for commands here:
https://github.com/spf13/cobra/blob/v1.2.1/command.go#L946-L950
then `PreRun` is only called later, after the command has already been resolved:
https://github.com/spf13/cobra/blob/v1.2.1/command.go#L974
https://github.com/spf13/cobra/blob/v1.2.1/command.go#L849

If I try a simplified experiment like this:
```
PreRun: func(cmd *cobra.Command, args []string) {cmd.SetArgs([]string{"sync"}))
```
The `PreRun` function does run when I execute the application without a command, but the inserted argument doesn't have any effect. This must be done before the call to `rootCmd.Execute()`.

This code was based on some discussions on the subject in the Cobra repo:

- https://github.com/spf13/cobra/issues/725
- https://github.com/spf13/cobra/issues/823

It seems this was the approach settled on.

Reply:
Ok then, I wasn't sure it was feasible in any case.